presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java

```diff
@@ -0,0 +1,305 @@
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestRetryDriver
+{
+    private static final Duration MIN_DURATION = new Duration(10, MILLISECONDS);
+    private static final Duration MAX_DURATION = new Duration(20, MILLISECONDS);
+    private static final Duration MAX_RETRY_TIME = new Duration(60, SECONDS);
+    private static final double BACK_OFF_SCALE_FACTOR = 1.5;
+    private static final int MAX_RETRY_ATTEMPTS = 10;
+    private static final Integer RESULT = 123;
+
+    private final ListeningScheduledExecutorService retryExecutor;
+    private final RetryDriver retry;
+
+    public TestRetryDriver()
+    {
+        retryExecutor = MoreExecutors.listeningDecorator(newScheduledThreadPool(4, threadsNamed("test-retry-retry-%s")));
+        this.retry = RetryDriver.retry(retryExecutor)
+                .exponentialBackoff(MIN_DURATION, MAX_DURATION, MAX_RETRY_TIME, BACK_OFF_SCALE_FACTOR)
+                .maxAttempts(MAX_RETRY_ATTEMPTS);
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void close()
+    {
+        retryExecutor.shutdownNow();
+    }
+
+    enum AttemptStatus
+    {
+        SUCCESS,
+        ASYNC_FAILURE,
+        SYNC_FAILURE,
+    }
+
+    private static class TaskTracker
+    {
+        private final List<AttemptStatus> attempts;
+        private final AtomicInteger nextAttempt;
+
+        public TaskTracker(AttemptStatus... attempts)
+        {
+            this.attempts = ImmutableList.copyOf(attempts);
+            this.nextAttempt = new AtomicInteger(0);
+        }
+
+        public int totalAttemptsMade()
+        {
+            return nextAttempt.get();
+        }
+
+        public AttemptStatus getNextAttemptStatus()
+        {
+            return attempts.get(nextAttempt.getAndIncrement());
+        }
+    }
+
+    private static class SyncTaskImpl
+            implements Callable<Integer>
+    {
+        private final TaskTracker taskTracker;
+
+        public SyncTaskImpl(AttemptStatus... attempts)
+        {
+            taskTracker = new TaskTracker(attempts);
+        }
+
+        public int totalAttemptsMade()
+        {
+            return taskTracker.totalAttemptsMade();
+        }
+
+        @Override
+        public Integer call()
+        {
+            switch (taskTracker.getNextAttemptStatus()) {
+                case SUCCESS:
+                    return RESULT;
+                case SYNC_FAILURE:
+                    throw new SyncException("Failed on call " + taskTracker.totalAttemptsMade());
+                default:
+                    throw new RuntimeException("Unexpected exception type");
+            }
+        }
+    }
+
+    private static class AsyncTaskImpl
```

Comment:
Move all helper classes to the bottom of the class.

Reply:
Done.